### PR TITLE
Bloom player

### PR DESCRIPTION
**Pull request by nikkuAg**
## Summary
Added a new plugin for bloom format files and created a runner in the hashi to render bloom files in the iframe client.

Version 1:
https://github.com/user-attachments/assets/1d1d73f8-b685-4acd-a68a-0a03ae7657cc

Version 2:
https://github.com/user-attachments/assets/2eb5ffc8-b47e-42e5-9b96-22a0b2b58882

Version 3:

https://github.com/user-attachments/assets/035c554b-7158-4d20-911f-69d64a4ebe12

https://github.com/user-attachments/assets/46c15024-adb6-402d-b068-ab60e7121f11

…

## References
#12237 
[#36](https://github.com/learningequality/kolibri-ecosystem/issues/36)
#12531
…

## Reviewer guidance

…

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -85,6 +85,7 @@ js-dist
 !kolibri/**/static/assets/
 !kolibri/**/static/assets/*
 !kolibri/**/static/assets/**/*
+!kolibri/**/static/bloom/
```
Comment (rtibbles) [inline .gitignore]: Let's make this more specific (like the h5p one below) and put it alongside the h5p one, just as they are thematically related.

File: kolibri/core/content/static/bloom/bloomplayer.htm [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+<!DOCTYPE html>
+<html>
+    <head>
+        <meta charset="UTF-8" />
+    </head>
+    <body style="background-color: #2e2e2e;">
+        <div id="root"><span style="color: #d65649">Loading Bloom Player...</span></div>
+    <script type="text/javascript" src="bloomPlayer-469de4b6cca7dc5c3e04.min.js"></script></body>
+    <!-- At build time, we insert a script tag pointing at bloomPlayer.min.js but with a hash created at build time,
```
Comment (rtibbles) [inline kolibri/core/content/static/bloom/bloomplayer.htm]: This comment seems at odds with the script tag above?

Where is the CSS loaded from?

File: kolibri/core/content/static/bloom/bloomplayer.htm [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+        to avoid getting a stale version from a cache (while allowing us to cache it for a long time).
+        It has a style sheet which sets the same color on the body, but it takes a while to fetch and load,
+        so we reduce flicker by setting a background color explicitly. -->
+</html>
```
Comment (rtibbles) [inline kolibri/core/content/static/bloom/bloomplayer.htm:13]: I can't comment on the binary files below - but are the mp3 files referenced in the JS?

File: packages/hashi/package.json [hunk 4]
```diff
@@ -13,14 +13,15 @@
...
+    "copy-webpack-plugin": "^12.0.2",
     "eslint-plugin-compat": "^6.0.0",
     "html-webpack-plugin": "5.6.0",
     "jquery": "3.5.1",
     "mutationobserver-shim": "^0.3.7",
     "purgecss": "^6.0.0"
   },
   "dependencies": {
-    "core-js": "3.38",
+    "core-js": "3.37",
```
Comment (rtibbles) [inline packages/hashi/package.json]: Not sure why this has been downgraded?

File: packages/hashi/src/Bloom/BloomRunner.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    // For each Bloom package, an object that maps the original file
+    // reference in the Bloom file to a blob URL reference
+    this.packageFiles = {};
+    // The iframe that we should be loading Bloom in - this is probably not the
+    // same as the current window context that the Bloom constructor has been
+    // invoked in.
+    this.iframe = iframe;
+    // This is the path to the Bloom file which we load in its entirety.
+    this.filepath = filepath;
+    // A fallback URL to the zipcontent endpoint for this H5P file
```
Comment (rtibbles) [inline packages/hashi/src/Bloom/BloomRunner.js:41]: *Bloompub file.

File: packages/hashi/src/Bloom/BloomRunner.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    });
+  }
+
+  /*
+   * Start Bloom player in the contentWindow.
+   */
+  initBloom() {
+    try {
+      this.loaded();
+      this.iframe.src = `../bloom/bloomplayer.htm?url=${this.contentUrl}&distributionUrl=${this.distributionUrl}&metaJsonUrl=${this.metaUrl}`;
```
Comment (rtibbles) [inline packages/hashi/src/Bloom/BloomRunner.js]: Makes sense!

File: packages/kolibri-zip/src/fileUtils.js [hunk 7]
```diff
@@ -127,4 +223,5 @@ export const defaultFilePathMappers = {
   htm: DOMMapper,
   xhtml: DOMMapper,
   xml: DOMMapper,
+  bloom: BloomMapper,
```
Comment (rtibbles) [inline packages/kolibri-zip/src/fileUtils.js]: Please reference back to my comments on your previous PR - you have defined a mapper for a bloom filetype, but that is not a filetype that is found inside the archive file.

Instead, this work should be done to enhance the DOMMapper class.

File: packages/hashi/package.json [hunk 8]
```diff
@@ -13,14 +13,15 @@
   "author": "Learning Equality",
   "license": "MIT",
   "devDependencies": {
+    "copy-webpack-plugin": "^12.0.2",
```
Comment (rtibbles) [inline packages/hashi/package.json]: This doesn't seem to be used at all now?

**Review by rtibbles — CHANGES_REQUESTED** [review]
Some things from the previous review have been addressed: https://github.com/learningequality/kolibri/pull/12531#pullrequestreview-2248964893 but the way the mapper is being referenced and applied is not quite fitting in with how kolibri-zip operates. A couple of other questions/clarifications.

File: kolibri/core/content/static/bloom/bloomplayer.htm (hunk 2 above)
Comment (nikkuAg) [reply kolibri/core/content/static/bloom/bloomplayer.htm]: CSS are loaded using the DOMWrapper

File: kolibri/core/content/static/bloom/bloomplayer.htm (hunk 3 above)
Comment (nikkuAg) [reply kolibri/core/content/static/bloom/bloomplayer.htm:13]: Yes, mp3 files are being used in js files to play audio for correct and wrong answers

**Comment by nikkuAg** [comment]
Now we can save the progress for bloom format files using the pages viewed
https://github.com/user-attachments/assets/6915805d-e668-4712-b171-8b92b9abc978

**Comment by rtibbles** [comment]
This currently adds no new JS dependencies, so I would suggest reverting any changes to the `yarn.lock` file - this will also mean that you don't have conflicts any more.

**Review by rtibbles — COMMENTED** [review]
I think we may still be able to do some cleanup on this, but as long as this is working as intended, this should be good to merge.

**Comment by pcenov** [comment]
Hi @rtibbles - the player in its current implementation is functioning in a sense that I was able to open and play the downloaded bloompub files.
As potential follow-up issues I would mention the following:
1. Missing table of contents menu
2. Missing text decrease/increase and theme color settings
3. Missing search option
4. The fullscreen option within the player is duplicating the Enter fullscreen option above it
5. The selection effect of the left and right arrows is unnecessary large
6. There's no option to view 2 pages side by side
7. When comparing a bloompub file with the corresponding epub file sometimes the images or parts of the text are missing

**Comment by rtibbles** [comment]
Hi @pcenov,

I am guessing some of this is by comparison against the EPub renderer in Kolibri - in chatting with @nikkuAg, it seems that some of these may be limitations of the underlying Bloom Player software (although we may be able to address them).

This seems to be something that shows on a per book basis and has to be defined for the specific book, we think:
1. Missing table of contents menu

It doesn't seem that Bloom Player currently supports these, so would have to be filed as potential enhancements for the future:
2. Missing text decrease/increase and theme color settings
3. Missing search option
6. There's no option to view 2 pages side by side

We can make follow up issues to fix these in Kolibri:
4. The fullscreen option within the player is duplicating the Enter fullscreen option above it (this is directly configurable in the Bloom Player, so we can just turn it off there)
7. When comparing a bloompub file with the corresponding epub file sometimes the images or parts of the text are missing

This would probably need to be fixed in our fork of Bloom Player, but we can track the issue in Kolibri:
5. The selection effect of the left and right arrows is unnecessary large

**Review by rtibbles — APPROVED** [review]
All follow up issues filed, merging.

**Pull request merged**